rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,500 @@
+well as preventing a new version and an old version from operating on the same keys with different schema. This is
+cumbersome and makes the upgrade process a pain point for cluster admins. Scaling Auth down and back up also results in
+an [uneven load](https://github.com/gravitational/teleport/issues/7029) which can cause connectivity issues and backend
+latency that can result in a thundering herd.
+
+Migrations are applied during Auth initialization and prevent the instance from becoming ready and able to serve
+requests which causes downtime. To date, migrations have largely been applied in place on an existing key range, even
+for breaking changes, which prevents an older versions of Auth from operating on the data. This can result in data loss
+as new and older versions try to write different versions of a resource to the same key or worse, rendering older
+versions of Teleport unable to comprehend the data stored in the backend.
+
+## Details
+
+Instead of all migrations being a one-shot operation which break compatibility with previous versions of Auth we can
+leverage a phased approach which separates backend changes from application logic changes to allow Auth vN and Auth vN-1
+to coexist. This approach is, however, likely best suited for Cloud where we control the cadence and order of Teleport
+releases that are applied. Self-hosted customers should still follow the traditional upgrade procedure.
+
+To prevent having to split the changes across separate commits and ensure that they land in the correct release, we can
+leverage feature flags to allow all the changes to land at once and toggle the behavior when appropriate. Once the
+migration has been completed, the feature flags and old code can be removed. For self-hosted customers migrations should
+still only happen during a major release, which means major/breaking changes should not be backported.
+
+Any Cloud tenants which are lagging behind the current release and are required to jump several versions to land on the
+current release should revert to using recreate instead of a rolling deploy. We can also add a flag that indicates
+migrations should be performed during initialization instead of in the background like during the phased approach. While
+this will negate the changes described in this RFD to eliminate downtime, this kind of jump in releases should be the
+exception and not the norm.
+
+### Major/Breaking changes
+
+For major breaking changes (i.e., splitting a resource into multiple resources), resources must be migrated to a new
+key. For example, to migrate the data stored in `some/path/to/resource/<ID>` we must leave the original value at
+`some/path/to/resource/<ID>` as is, and write the migrated value to `some/new/path/to/resource/<ID>`. Keys should be
+versioned to determine which version of a resource exists at any given key range. For example, if nodes were migrated
+from `/nodes/default/<UUID>` it should be to `/nodes/v2/default/<UUID>`.
```

Comment:
```suggestion
from `/nodes/default/<UUID>` it should be to `/nodes_v2/default/<UUID>`.
```
```suggestion
from `/nodes/default/<UUID>` it should be to `/nodesV2/default/<UUID>`.
```

Let's just never reuse prefixes even if it would "surely" work.